mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h

```diff
@@ -0,0 +1,421 @@
+  MemoryAccessMatrix(const MemoryAccessMatrix &) = default;
+
+  /// Construct a matrix with the specified number of rows and columns.
+  MemoryAccessMatrix(unsigned nRows, unsigned nColumns);
+
+  /// Construct a matrix from a initializer list.
+  MemoryAccessMatrix(
+      std::initializer_list<std::initializer_list<Value>> initList);
+
+  /// Access the element at the specified \p row and \p column.
+  Value &at(unsigned row, unsigned column) {
+    assert(row < nRows && "Row outside of range");
+    assert(column < nColumns && "Column outside of range");
+    return data[row * nColumns + column];
+  }
+
+  Value at(unsigned row, unsigned column) const {
+    assert(row < nRows && "Row outside of range");
+    assert(column < nColumns && "Column outside of range");
+    return data[row * nColumns + column];
+  }
+
+  Value &operator()(unsigned row, unsigned column) { return at(row, column); }
+  Value operator()(unsigned row, unsigned column) const {
+    return at(row, column);
+  }
+
+  /// Swap \p column with \p otherColumn.
+  void swapColumns(unsigned column, unsigned otherColumn);
+
+  /// Swap \p row with \p otherRow.
+  void swapRows(unsigned row, unsigned otherRow);
+
+  unsigned getNumRows() const { return nRows; }
+
+  unsigned getNumColumns() const { return nColumns; }
+
+  /// Get a copy of the specified \p row.
+  SmallVector<Value> getRow(unsigned row) const;
+
+  /// Set the specified \p row to \p elems.
+  void setRow(unsigned row, ArrayRef<Value> elems);
+
+  /// Fill \p row with the given value \p val.
+  void fillRow(unsigned row, Value val);
+
+  /// Add an extra row at the bottom of the matrix.
+  unsigned appendRow();
+
+  /// Add an extra row at the bottom of the matrix and copy the given elements
+  /// \p elems into the new row.
+  unsigned appendRow(ArrayRef<Value> elems);
+
+  /// Get a copy of the specified \p column.
+  SmallVector<Value> getColumn(unsigned column) const;
+
+  /// Set the specified \p column to \p elems.
+  void setColumn(unsigned col, ArrayRef<Value> elems);
+
+  /// Fill \p col with the given value \p val.
+  void fillColumn(unsigned col, Value val);
+
+  /// Fill the matrix with the given value \p val.
+  void fill(Value val);
+
+  /// Construct a new matrix containing the specified \p rows.
+  MemoryAccessMatrix getRows(std::set<unsigned> rows) const;
+
+  /// Construct a new matrix containing the specified \p columns.
+  MemoryAccessMatrix getColumns(std::set<unsigned> columns) const;
+
+  /// Construct a new matrix containing the sub-matrix specified by \p rows and
+  /// \p columns.
+  MemoryAccessMatrix getSubMatrix(std::set<unsigned> rows,
+                                  std::set<unsigned> columns) const;
+
+  //===----------------------------------------------------------------------===//
+  // Shape Queries
+  //===----------------------------------------------------------------------===//
+
+  /// Returns true if the matrix has equal number of rows and columns.
```

Comment:
nit: as the previous comments uses `Construct`, `Add`, `Get` without `s`, let's change all these `Returns` to `Return`.